iwyu_ast_util.cc

```diff
@@ -546,32 +560,29 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
 
   // --- Methods on RecursiveASTVisitor
   bool TraverseType(QualType type) {
-    TraverseArgumentsOfSugaredTemplates(type);
-    return Base::TraverseType(type);
+    return TraverseTypeHelper(type);
   }
 
   bool TraverseTypeLoc(TypeLoc type_loc) {
-    TraverseArgumentsOfSugaredTemplates(type_loc.getType());
-    return Base::TraverseTypeLoc(type_loc);
-  }
-
-  bool VisitType(Type* type) {
-    seen_types_.insert(type);
-    return true;
+    return TraverseTypeHelper(type_loc.getType());
   }
 
  private:
-  // Clang doesn't traverse underlying type for most of the sugar types.
-  // If a TemplateSpecializationType occurs in a sugaring chain, traverse its
-  // arguments explicitly, otherwise they aren't put into resugar_map.
-  void TraverseArgumentsOfSugaredTemplates(QualType type) {
-    if (type.isNull())
-      return;
-
-    if (const auto* template_spec = type->getAs<TemplateSpecializationType>()) {
-      for (const TemplateArgument& arg : template_spec->template_arguments())
-        TraverseTemplateArgument(arg);
-    }
+  bool TraverseTypeHelper(QualType qual_type) {
+    if (qual_type.isNull())
+      return false;
```

Comment:
I'm not sure about `false`. Maybe, empty type node should not stop the traversal? Or let the base class implementation decide?

Reply:
That's what `iwyu.cc` `TraverseType` does:
```
  bool TraverseType(QualType qualtype) {
    if (qualtype.isNull())
      return Base::TraverseType(qualtype);
```

Whether it's correct or not, I think it would make sense to be consistent.

Reply:
Thanks!